crates/m3/src/gadgets/div.rs

```diff
@@ -0,0 +1,414 @@
+// Copyright 2025 Irreducible Inc.
+
+use std::array;
+
+use binius_field::{packed::set_packed_slice, Field, PackedExtension, PackedField};
+use itertools::izip;
+
+use crate::{
+	builder::{Col, TableBuilder, TableWitnessSegment, B1, B128, B32, B64},
+	gadgets::{
+		mul::{MulSS32, MulUU32, SignConverter, UnsignedMulPrimitives},
+		u32::{add::WideAdd, sub::WideSub, U32AddFlags, U32SubFlags},
+		util::pack_fp,
+	},
+};
+
+/// Gadget for unsigned division of two u32s.
+///
+/// `p = q*a + r`
+#[derive(Debug)]
+pub struct DivUU32 {
+	mul_inner: MulUU32,
+	sum: WideAdd<u64, 64>,
+	sub: WideSub<u64, 64>,
+
+	p_in_bits: [Col<B1>; 32],
+	q_in_bits: [Col<B1>; 32],
+	out_div_bits: [Col<B1>; 32],
+	out_rem_bits: [Col<B1>; 32],
+
+	pub p_in: Col<B32>,
+	pub q_in: Col<B32>,
+	pub out_div: Col<B32>,
+	pub out_rem: Col<B32>,
+}
+
+impl DivUU32 {
+	#[allow(clippy::needless_range_loop)]
+	pub fn new(table: &mut TableBuilder) -> Self {
+		let zero = table.add_constant("zero", [B1::ZERO]);
+		let p_in_bits = table.add_committed_multiple("p_in_bits");
+		let q_in_bits = table.add_committed_multiple("q_in_bits");
+
+		let p_in = table.add_computed("p_in", pack_fp(p_in_bits));
+		let q_in = table.add_computed("q_in", pack_fp(q_in_bits));
+
+		let zero_extend_p: [_; 64] = array::from_fn(|i| if i < 32 { p_in_bits[i] } else { zero });
+
+		let out_div_bits = table.add_committed_multiple("out_div_bits");
+		let out_rem_bits = table.add_committed_multiple("out_rem_bits");
+
+		let zero_extend_q: [_; 64] = array::from_fn(|i| if i < 32 { q_in_bits[i] } else { zero });
+		let zero_extend_rem = array::from_fn(|i| if i < 32 { out_rem_bits[i] } else { zero });
+
+		let out_div = table.add_computed("out_div", pack_fp(out_div_bits));
+		let out_rem = table.add_computed("out_rem", pack_fp(out_rem_bits));
+
+		let mul_inner = MulUU32::with_inputs(table, q_in_bits, out_div_bits);
+
+		// Check q is non-zero
+		table.assert_nonzero(q_in);
+
+		let product_cols = array::from_fn(|i| {
+			if i < 32 {
+				mul_inner.out_low_bits[i]
+			} else {
+				mul_inner.out_high_bits[i - 32]
+			}
+		});
+
+		// Check p = q * a + r in 64 bits
+		let sum = WideAdd::<u64, 64>::new(
+			table,
+			product_cols,
+			zero_extend_rem,
+			U32AddFlags {
+				commit_zout: true,
+				expose_final_carry: true,
+				..Default::default()
+			},
+		);
+
+		for bit in 0..64 {
+			table.assert_zero(
+				format!("division_satisfied_{bit}"),
```

Comment:
Just to be consistent with index notation we use elsewhere.

```suggestion
				format!("division_satisfied[{bit}]"),
```